File: leetcode/301-400/308-range-sum-query-2d-mutable/bench/fenwick2d.py
# Benchmark mirror — LeetCode 308, Range Sum Query 2D (Mutable).
# Same 2D Fenwick tree, same LCG-generated operation script, same masked sink
# as fenwick2d.kara. See ../README.md § Benchmarks.
#
# The sink masks rather than taking a modulus: Python's `%` FLOORS while
# C/Rust/Go/kara truncate toward zero, so a modulo sink over a signed running
# total prints a different number here than in every other mirror (measured on
# #303). Masking is two's-complement in all five languages.


def main():
    n = 256
    stride = n + 1
    ops = 100000
    passes = 54

    tree = [0] * ((n + 1) * stride)
    data = [0] * (n * n)
    kind = [0] * ops
    o1 = [0] * ops
    o2 = [0] * ops
    o3 = [0] * ops
    o4 = [0] * ops

    state = 20308
    for k in range(ops):
        state = (state * 1103515245 + 12345) % 2147483648
        t = state % 2
        state = (state * 1103515245 + 12345) % 2147483648
        a = state % n
        state = (state * 1103515245 + 12345) % 2147483648
        b = state % n
        state = (state * 1103515245 + 12345) % 2147483648
        c = state % n
        state = (state * 1103515245 + 12345) % 2147483648
        d = state % n
        kind[k] = t
        if t == 0:
            o1[k], o2[k], o3[k], o4[k] = a, b, c % 2001 - 1000, 0
        else:
            o1[k], o3[k] = (a, c) if a <= c else (c, a)
            o2[k], o4[k] = (b, d) if b <= d else (d, b)

    checksum = 0
    for _ in range(passes):
        for k in range(ops):
            if kind[k] == 0:
                r, c = o1[k], o2[k]
                delta = o3[k] - data[r * n + c]
                data[r * n + c] = o3[k]
                x = r + 1
                while x <= n:
                    y = c + 1
                    while y <= n:
                        tree[x * stride + y] += delta
                        y += y & -y
                    x += x & -x
            else:
                r1, c1, r2, c2 = o1[k], o2[k], o3[k] + 1, o4[k] + 1
                total = 0
                for qi in range(4):
                    px, py, sign = r2, c2, 1
                    if qi == 1:
                        px, sign = r1, -1
                    if qi == 2:
                        py, sign = c1, -1
                    if qi == 3:
                        px, py = r1, c1
                    sub = 0
                    x = px
                    while x > 0:
                        y = py
                        while y > 0:
                            sub += tree[x * stride + y]
                            y -= y & -y
                        x -= x & -x
                    total += sign * sub
                checksum = (checksum + total) & 0x3FFFFFFF
    print(f"checksum {checksum}")


main()

File: leetcode/301-400/308-range-sum-query-2d-mutable/bench/fenwick2d.rs
// Benchmark mirror — LeetCode 308, Range Sum Query 2D (Mutable).
// Same 2D Fenwick tree, same LCG-generated operation script, same masked sink
// as fenwick2d.kara. See ../README.md § Benchmarks.

fn main() {
    let n: i64 = 256;
    let stride: i64 = n + 1;
    let ops: i64 = 100000;
    let passes: i64 = 54;

    let mut tree: Vec<i64> = vec![0; ((n + 1) * stride) as usize];
    let mut data: Vec<i64> = vec![0; (n * n) as usize];
    let mut kind: Vec<i64> = Vec::with_capacity(ops as usize);
    let mut o1: Vec<i64> = Vec::with_capacity(ops as usize);
    let mut o2: Vec<i64> = Vec::with_capacity(ops as usize);
    let mut o3: Vec<i64> = Vec::with_capacity(ops as usize);
    let mut o4: Vec<i64> = Vec::with_capacity(ops as usize);

    let mut state: i64 = 20308;
    for _ in 0..ops {
        state = (state * 1103515245 + 12345) % 2147483648; let t = state % 2;
        state = (state * 1103515245 + 12345) % 2147483648; let a = state % n;
        state = (state * 1103515245 + 12345) % 2147483648; let b = state % n;
        state = (state * 1103515245 + 12345) % 2147483648; let c = state % n;
        state = (state * 1103515245 + 12345) % 2147483648; let d = state % n;
        kind.push(t);
        if t == 0 {
            o1.push(a); o2.push(b); o3.push(c % 2001 - 1000); o4.push(0);
        } else {
            if a <= c { o1.push(a); o3.push(c); } else { o1.push(c); o3.push(a); }
            if b <= d { o2.push(b); o4.push(d); } else { o2.push(d); o4.push(b); }
        }
    }

    let mut checksum: i64 = 0;
    for _ in 0..passes {
        for k in 0..ops as usize {
            if kind[k] == 0 {
                let r = o1[k];
                let c = o2[k];
                let delta = o3[k] - data[(r * n + c) as usize];
                data[(r * n + c) as usize] = o3[k];
                let mut x = r + 1;
                while x <= n {
                    let mut y = c + 1;
                    while y <= n {
                        tree[(x * stride + y) as usize] += delta;
                        y += y & -y;
                    }
                    x += x & -x;
                }
            } else {
                let r1 = o1[k];
                let c1 = o2[k];
                let r2 = o3[k] + 1;
                let c2 = o4[k] + 1;
                let mut total: i64 = 0;
                for qi in 0..4 {
                    let mut px = r2;
                    let mut py = c2;
                    let mut sign: i64 = 1;
                    if qi == 1 { px = r1; sign = -1; }
                    if qi == 2 { py = c1; sign = -1; }
                    if qi == 3 { px = r1; py = c1; }
                    let mut sub: i64 = 0;
                    let mut x = px;
                    while x > 0 {
                        let mut y = py;
                        while y > 0 {
                            sub += tree[(x * stride + y) as usize];
                            y -= y & -y;
                        }
                        x -= x & -x;
                    }
                    total += sign * sub;
                }
                checksum = (checksum + total) & 0x3FFFFFFF;
            }
        }
    }
    println!("checksum {}", checksum);
}
